content/graphql/federation.md

```diff
@@ -0,0 +1,225 @@
+@Resolver("User")
+export class UsersResolvers {
+  constructor(private readonly usersService: UsersService) {}
+
+  @Query()
+  getUser(@Args("id") id: string) {
+    return this.usersService.findById(id);
+  }
+
+  @ResolveReference()
+  resolveReference(reference: { __typename: string; id: string }) {
+    return this.usersService.findById(reference.id);
+  }
+}
+```
+
+Finally, we hook it everything up in a module together with a `GraphQLFederationModule`. This module accepts the same options as the regular `GraphQLModule`.
+
+```typescript
+import { Module } from "@nestjs/common";
+import { GraphQLFederationModule } from "@nestjs/graphql";
+import { UsersResolvers } from "./users.resolvers";
+
+@Module({
+  imports: [
+    GraphQLFederationModule.forRoot({
+      typePaths: ["**/*.graphql"]
+    })
+  ],
+  providers: [UsersResolvers]
+})
+export class AppModule {}
+```
+
+#### Federated example: Posts
+
+The Posts service references the User type in its schema by marking it with the `extend` keyword. It also adds one property to the User type. Note the `@key` directive used for matching instances of User, and the `@external` directive indicating that the `id` field is managed elsewhere.
+
+```java
+type Post @key(fields: "id") {
+  id: ID!
+  title: String!
+  body: String!
+  user: User
+}
+
+extend type User @key(fields: "id") {
+  id: ID! @external
+  posts: [Post]
+}
+
+extend type Query {
+  getPosts: [Post]
+}
+```
+
+Our resolver has one method of interest here: `getUser`. It returns a reference containing `__typename` and any additional properties your application needs to resolve the reference, in this case only an `id`. The `__typename` is used by the GraphQL Gateway to pinpoint the microservice responsible for the User type and request the instance; The User service discussed above will be called on the `resolveReference` method.
+
+```typescript
+import { Query, Resolver, Parent, ResolveProperty } from "@nestjs/graphql";
+import { PostsService } from "./posts.service";
+import { Post } from "./posts.interfaces";
+
+@Resolver("Post")
+export class PostsResolvers {
+  constructor(private readonly postsService: PostsService) {}
+
+  @Query("getPosts")
+  getPosts() {
+    return this.postsService.findAll();
+  }
+
+  @ResolveProperty("user")
+  getUser(@Parent() post: Post) {
+    return { __typename: "User", id: post.userId };
+  }
+}
+```
+
+The Posts service has virtually the same module, but is included below for the sake of completeness
+
+```typescript
+import { Module } from "@nestjs/common";
+import { GraphQLFederationModule } from "@nestjs/graphql";
+import { PostsResolvers } from "./posts.resolvers";
+
+@Module({
+  imports: [
+    GraphQLFederationModule.forRoot({
+      typePaths: ["**/*.graphql"]
+    })
+  ],
+  providers: [PostsResolvers]
+})
+export class AppModule {}
+```
+
+#### Federated example: Gateway
+
+First make sure you install the optional dependency for the gateway: `npm install --save @apollo/gateway`.
```

Comment:
First make sure you install the optional dependency for the gateway: `npm install --save @apollo/gateway`.
->
First install the optional dependency for the gateway: 
```bash
npm install --save @apollo/gateway
```